src/secret_service.rs

```diff
@@ -380,28 +387,33 @@ impl SsCredential {
         require_unique: bool,
     ) -> Result<Vec<T>>
     where
-        F: Fn(&Item) -> Result<T>,
+        F: Fn(&Item<'_>) -> Result<T>,
         T: Sized,
     {
         let collection = ss.get_default_collection().map_err(decode_error)?;
         let attributes = self.search_attributes(true);
-        let search = collection.search_items(attributes).map_err(decode_error)?;
-        if require_unique {
-            if search.is_empty() && require_unique {
-                return Err(ErrorCode::NoEntry);
-            } else if search.len() > 1 {
-                let mut creds: Vec<Box<Credential>> = vec![];
-                for item in search.iter() {
-                    let cred = Self::new_from_item(item)?;
-                    creds.push(Box::new(cred))
-                }
-                return Err(ErrorCode::Ambiguous(creds));
+        let items = collection.search_items(attributes).map_err(decode_error)?;
+
+        match items.len() {
+            0 => return Err(ErrorCode::NoEntry),
+            n if n > 1 && require_unique => {
+                let creds: Result<Vec<Box<Credential>>> = items
+                    .iter()
+                    .map(|item| {
+                        Self::new_from_item(item).map(|cred| Box::new(cred) as Box<Credential>)
+                    })
+                    .collect();
+                return Err(ErrorCode::Ambiguous(creds?));
             }
+            _ => {}
         }
-        let mut results: Vec<T> = vec![];
-        for item in search.iter() {
+
+        let mut results: Vec<T> = Vec::with_capacity(items.len());
+
+        for item in items.iter() {
             results.push(f(item)?);
         }
+
```

Comment:
same comment as above: this restructure is wrong (and not needed). The only difference between the code here and that above is that we are doing a collection search so the results are never locked.

Reply:
reverted